README.md

```diff
@@ -238,76 +238,119 @@ the DRAM implementation. However, here as well we will encounter are a few thing
+  The `.mut` will eventually be removed once the file is fully written. While it is not necessary it makes it a
+  bit safer to get rid of incomplete files when recovering from a failure.
+- `<uuid>.feldera`: Batch files used by various spines of a circuit.
+- `<uuid>/`: Directories that correspond to a checkpoint and have checkpoint specific files (note that a batch
+  may be used by many checkpoints, hence they live in the base directory).
+- `<uuid>/pspine-batches-<persistent-id>.dat`: There is one for every spine in the circuit. These are rkyv files
+  of type `Vec<String>` and they contain just a list of batch files in by the spine for this given checkpoint.
+- `<uuid>/psine-<persistent-id>.dat`: This contains a rkyv representation of all the state necessary to completely
+  reconstruct `struct Spine`.
+
+Q: Why do we have two set of files, `pspine-*.dat` and `pspine-batches-*.dat` for spine meta-data: It is easier to 
+read `pspine-batches-*.dat` from anywhere in the code as it doesn't require me to know the exact generic type of \
+the Spine in order to read/write it with rkyv.
+
+#### Taking a checkpoint
+
+We add a `commit` to the circuit (`DBSPHandle`) to take a new checkpoint. The commit is named using an `uuid`
+and will lead to a `uuid` directory being created in the storage location for storing files
+specific to the checkpoint. We take a checkpoint by allowing the circuit handle to send 
+a checkpoint command to all workers, which then will invoke a `checkpoint` function on every node/operator. 
+The operators can indepdently have custom logic to persist state and/or reply with an error.
+The checkpoint command will wait synchronously until all workers have replied. If everyone
+was successful it will write a new entry for this checkpoint into `checkpoints.feldera.mut` and atomically
+rename it to `checkpoints.feldera`, which is the point at which the checkpoint was successfully
+created and is discoverable by a new process/circuit after.
+
+One thing to consider is that we do not want the user to invoke a checkpoint while the circuit is performing a
+step. But this is easy to ensure in rust as `step` is a mutable operation on the circuit.
+
+The most important operator to implement commit is the Spine. Since the Spine is roughly a set
+of batches, the commit implementation has to write this list out in a persistent way so that we can
+reinitialize the Spine again with the same set of files in case we resume with the circuit in a new process. 
+Potentially, there can be many Spines in a circuit, so we also need to name them so we can find them again. 
+For that we add persistent-ids for operators that are based on the global node-ids of the circuit.
+
+Another potential pitfall could be that batch-files are not fully written out to disk when we commit
+the file. Currently we flush (too often) whenever we completed writing a batch. In the future we would need
```

Comment:
completed -> complete
you can remove the phrase about the pitfall, this is clear enough